File: backend/app/schemas/video.py
from pydantic import BaseModel, Field


class VideoPageRequest(BaseModel):
    channel_url: str = Field(..., min_length=1)
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)
    enrich: bool = True
    include_total: bool = True


class VideoMetadata(BaseModel):
    video_id: str
    url: str
    title: str
    description: str | None = None
    thumbnail_url: str | None = None
    duration_seconds: int | None = None
    duration_text: str | None = None
    upload_date: str | None = None
    view_count: int | None = None
    like_count: int | None = None
    comment_count: int | None = None
    channel_id: str | None = None
    channel_name: str | None = None
    channel_url: str | None = None
    tags: list[str] = []
    availability: str | None = None
    is_short: bool = False

    def model_post_init(self, __context):
        # YouTube Shorts are typically ≤ 60 seconds
        if self.duration_seconds is not None:
            self.is_short = self.duration_seconds <= 60


class VideoPageResponse(BaseModel):
    channel_url: str
    page: int
    page_size: int
    count: int
    total_count: int | None = None
    has_more: bool
    videos: list[VideoMetadata]


class VideoSearchRequest(BaseModel):
    channel_url: str = Field(..., min_length=1)
    query: str = Field(..., min_length=1)
    limit: int | None = Field(default=None, ge=1, le=1000)
    enrich: bool = True
    auto_select_threshold: float = Field(default=55, ge=0, le=100)


class TopicVideoSearchRequest(BaseModel):
    query: str = Field(..., min_length=1)
    limit: int = Field(default=50, ge=1, le=100)
    enrich: bool = False
    auto_select_threshold: float = Field(default=80, ge=0, le=100)


class RankVideosRequest(BaseModel):
    query: str = Field(..., min_length=1)
    videos: list[VideoMetadata]
    auto_select_threshold: float = Field(default=55, ge=0, le=100)


class RankedVideo(VideoMetadata):
    topic_score: float
    selected: bool
    rank_reason: str


class RankVideosResponse(BaseModel):
    query: str
    auto_select_threshold: float
    videos: list[RankedVideo]


class VideoSearchResponse(BaseModel):
    channel_url: str
    query: str
    total_count: int | None = None
    candidate_count: int
    videos: list[RankedVideo]


class TopicVideoSearchResponse(BaseModel):
    query: str
    total_count: int | None = None
    candidate_count: int
    videos: list[RankedVideo]


class KnowledgeBaseVideoInput(BaseModel):
    video_id: str
    url: str
    title: str | None = None


class KnowledgeBaseCreateRequest(BaseModel):
    channel_name: str = Field(..., min_length=1)
    channel_url: str = Field(..., min_length=1)
    output_dir: str = Field(default="", min_length=0)
    videos: list[KnowledgeBaseVideoInput] = Field(..., min_length=1)
    include_comments: bool = True
    max_comments: int = Field(default=50, ge=0, le=500)
    file_per_video: bool = False
    download_transcribe_if_missing: bool = False


class KnowledgeBaseFileResult(BaseModel):
    video_id: str
    title: str
    file_path: str
    transcript_status: str
    comments_count: int


class KnowledgeBaseTranscriptSummary(BaseModel):
    available: int = 0
    blocked_by_youtube: int = 0
    unavailable: int = 0


class KnowledgeBaseCreateResponse(BaseModel):
    output_path: str
    count: int
    files: list[KnowledgeBaseFileResult]
    warnings: list[str] = []
    transcript_summary: KnowledgeBaseTranscriptSummary = KnowledgeBaseTranscriptSummary()
    download_url: str | None = None
    download_filename: str | None = None


class FolderPickResponse(BaseModel):
    path: str | None = None
    cancelled: bool = False

File: frontend/src/types/video.ts
export type VideoMetadata = {
  video_id: string;
  url: string;
  title: string;
  description: string | null;
  thumbnail_url: string | null;
  duration_seconds: number | null;
  duration_text: string | null;
  upload_date: string | null;
  view_count: number | null;
  like_count: number | null;
  comment_count: number | null;
  channel_id: string | null;
  channel_name: string | null;
  channel_url: string | null;
  tags: string[];
  availability: string | null;
  is_short: boolean;
};

export type VideoPageResponse = {
  channel_url: string;
  page: number;
  page_size: number;
  count: number;
  total_count: number | null;
  has_more: boolean;
  videos: VideoMetadata[];
};

export type RankedVideo = VideoMetadata & {
  topic_score: number;
  selected: boolean;
  rank_reason: string;
};

export type VideoSearchResponse = {
  channel_url: string;
  query: string;
  total_count: number | null;
  candidate_count: number;
  videos: RankedVideo[];
};

export type TopicVideoSearchResponse = {
  query: string;
  total_count: number | null;
  candidate_count: number;
  videos: RankedVideo[];
};

export type KnowledgeBaseFileResult = {
  video_id: string;
  title: string;
  file_path: string;
  transcript_status: string;
  comments_count: number;
};

export type KnowledgeBaseCreateResponse = {
  output_path: string;
  count: number;
  files: KnowledgeBaseFileResult[];
  warnings: string[];
  transcript_summary: {
    available: number;
    blocked_by_youtube: number;
    unavailable: number;
  };
  download_url: string | null;
  download_filename: string | null;
};

export type KnowledgeBaseProgressEvent =
  | {
      type: "start";
      total: number;
      completed: number;
      output_path: string;
    }
  | {
      type: "video_started";
      index: number;
      total: number;
      completed: number;
      video: {
        video_id: string;
        title: string;
        url: string;
      };
    }
  | {
      type: "video_done";
      index: number;
      total: number;
      completed: number;
      video: KnowledgeBaseFileResult;
    }
  | {
      type: "video_error";
      index: number;
      total: number;
      completed: number;
      video: {
        video_id: string;
        title: string;
        url: string;
      };
      message: string;
    }
  | {
      type: "done";
      total: number;
      completed: number;
      result: KnowledgeBaseCreateResponse;
    };

export type FolderPickResponse = {
  path: string | null;
  cancelled: boolean;
};
